ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -30,16 +36,34 @@
 public class ByteBufferReadableStreamChannel implements ReadableStreamChannel {
   private final AtomicBoolean channelOpen = new AtomicBoolean(true);
   private final AtomicBoolean channelEmptied = new AtomicBoolean(false);
-  private final ByteBuffer buffer;
-  private final int size;
+  private final ReentrantLock contentLock = new ReentrantLock();
+  private final List<ByteBuffer> buffers;
+
+  private int size;
 
   /**
    * Constructs a {@link ReadableStreamChannel} whose read operations return data from the provided {@code buffer}.
    * @param buffer the {@link ByteBuffer} that is used to retrieve data from on invocation of read operations.
    */
   public ByteBufferReadableStreamChannel(ByteBuffer buffer) {
-    this.buffer = buffer;
-    size = buffer.remaining();
+    this(new ArrayList<>(Arrays.asList(buffer)));
+  }
+
+  /**
+   * Constructs a {@link ReadableStreamChannel} whose read operations return data from the provided {@code bufferList}
+   * and null {@link ByteBuffer} at the end of buffers signifies end of content.
+   * @param bufferList the {@link ByteBuffer} that is used to retrieve data from on invocation of read operations.
+   */
+
+  public ByteBufferReadableStreamChannel(List<ByteBuffer> bufferList) {
+    if (bufferList != null) {
```

Comment:
`bufferList` should not be `null` - you should throw a `IllegalArgumentException`